File: app-android/src/androidTest/java/com/ustadmobile/test/port/android/view/ContentEntryDetailEspressoTest.java
package com.ustadmobile.test.port.android.view;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.os.Environment;
import android.support.test.InstrumentationRegistry;
import android.support.test.espresso.contrib.RecyclerViewActions;
import android.support.test.espresso.intent.rule.IntentsTestRule;
import android.support.test.espresso.matcher.ViewMatchers;
import android.support.test.rule.GrantPermissionRule;
import android.support.test.runner.AndroidJUnit4;
import android.view.View;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.ContentEntryListPresenter;
import com.ustadmobile.core.db.JobStatus;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.UmLiveData;
import com.ustadmobile.core.db.dao.ContentEntryContentEntryFileJoinDao;
import com.ustadmobile.core.db.dao.ContentEntryDao;
import com.ustadmobile.core.db.dao.ContentEntryFileDao;
import com.ustadmobile.core.db.dao.ContentEntryFileStatusDao;
import com.ustadmobile.core.db.dao.ContentEntryRelatedEntryJoinDao;
import com.ustadmobile.core.db.dao.ContentEntryStatusDao;
import com.ustadmobile.core.db.dao.LanguageDao;
import com.ustadmobile.core.view.WebChunkView;
import com.ustadmobile.lib.db.entities.ContentEntry;
import com.ustadmobile.lib.db.entities.ContentEntryContentEntryFileJoin;
import com.ustadmobile.lib.db.entities.ContentEntryFile;
import com.ustadmobile.lib.db.entities.ContentEntryFileStatus;
import com.ustadmobile.lib.db.entities.ContentEntryRelatedEntryJoin;
import com.ustadmobile.lib.db.entities.ContentEntryStatus;
import com.ustadmobile.lib.db.entities.Language;
import com.ustadmobile.port.android.view.ContentEntryDetailActivity;
import com.ustadmobile.port.android.view.WebChunkActivity;

import org.hamcrest.Matchers;
import org.hamcrest.core.AllOf;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.intent.Intents.intended;
import static android.support.test.espresso.intent.matcher.IntentMatchers.hasComponent;
import static android.support.test.espresso.intent.matcher.IntentMatchers.hasExtra;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static com.ustadmobile.core.controller.ContentEntryDetailPresenter.ARG_CONTENT_ENTRY_UID;
import static com.ustadmobile.test.port.android.UmAndroidTestUtil.readFromTestResources;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.core.IsEqual.equalTo;

@RunWith(AndroidJUnit4.class)
public class ContentEntryDetailEspressoTest {

    @Rule
    public IntentsTestRule<ContentEntryDetailActivity> mActivityRule =
            new IntentsTestRule<>(ContentEntryDetailActivity.class, false, false);


    @Rule
    public GrantPermissionRule permissionRule =
            GrantPermissionRule.grant(Manifest.permission.WRITE_EXTERNAL_STORAGE,
                    Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.ACCESS_COARSE_LOCATION);


    private ContentEntryStatusDao statusDao;
    private File targetFile;

    public UmAppDatabase getDb() {
        Context context = InstrumentationRegistry.getTargetContext();
        UmAppDatabase db = UmAppDatabase.getInstance(context);
        db.clearAllTables();
        return db.getRepository("https://localhost", "");
    }

    public void createDummyContent() throws IOException {
        UmAppDatabase repo = getDb();

        ContentEntryDao contentDao = repo.getContentEntryDao();
        ContentEntryFileDao contentFileDao = repo.getContentEntryFileDao();
        ContentEntryContentEntryFileJoinDao contentEntryFileJoinDao = repo.getContentEntryContentEntryFileJoinDao();
        ContentEntryRelatedEntryJoinDao contentEntryRelatedEntryJoinDao = repo.getContentEntryRelatedEntryJoinDao();
        LanguageDao languageDao = repo.getLanguageDao();
        ContentEntryFileDao fileDao = repo.getContentEntryFileDao();

        UmAppDatabase app = UmAppDatabase.getInstance(null);
        statusDao = app.getContentEntryStatusDao();
        ContentEntryFileStatusDao fileStatusDao = app.getContentEntryFileStatusDao();

        Language englishLang = new Language();
        englishLang.setLangUid(1);
        englishLang.setName("English");
        englishLang.setIso_639_1_standard("en");
        englishLang.setIso_639_2_standard("eng");
        englishLang.setIso_639_3_standard("eng");
        languageDao.insert(englishLang);

        Language arabicLang = new Language();
        arabicLang.setLangUid(2);
        arabicLang.setName("Arabic");
        arabicLang.setIso_639_1_standard("ar");
        arabicLang.setIso_639_2_standard("ara");
        arabicLang.setIso_639_3_standard("ara");
        languageDao.insert(arabicLang);

        Language spanishLang = new Language();
        spanishLang.setLangUid(3);
        spanishLang.setName("Spanish");
        spanishLang.setIso_639_1_standard("es");
        spanishLang.setIso_639_2_standard("esp");
        spanishLang.setIso_639_3_standard("esp");
        languageDao.insert(spanishLang);


        ContentEntry quiz = new ContentEntry();
        quiz.setContentEntryUid(6);
        quiz.setTitle("Quiz Time");
        quiz.setThumbnailUrl("https://www.africanstorybook.org/img/asb120.png");
        quiz.setDescription("All content");
        quiz.setPublisher("CK12");
        quiz.setAuthor("Binge");
        quiz.setPrimaryLanguageUid(1);
        quiz.setLeaf(true);
        contentDao.insert(quiz);

        ContentEntryFile contentEntryFile = new ContentEntryFile();
        contentEntryFile.setMimeType("application/zip");
        contentEntryFile.setFileSize(10);
        contentEntryFile.setLastModified(1540728217);
        contentEntryFile.setContentEntryFileUid(8);
        contentFileDao.insert(contentEntryFile);

        ContentEntryContentEntryFileJoin fileJoin = new ContentEntryContentEntryFileJoin();
        fileJoin.setCecefjContentEntryFileUid(contentEntryFile.getContentEntryFileUid());
        fileJoin.setCecefjContentEntryUid(quiz.getContentEntryUid());
        fileJoin.setCecefjUid(9);
        contentEntryFileJoinDao.insert(fileJoin);

        // arabic
        ContentEntry arabicQuiz = new ContentEntry();
        arabicQuiz.setContentEntryUid(10);
        arabicQuiz.setTitle("وقت الاختبار");
        arabicQuiz.setThumbnailUrl("https://www.africanstorybook.org/img/asb120.png");
        arabicQuiz.setDescription("كل المحتوى");
        arabicQuiz.setPublisher("CK12");
        arabicQuiz.setAuthor("حفلة");
        arabicQuiz.setPrimaryLanguageUid(2);
        arabicQuiz.setLeaf(true);
        contentDao.insert(arabicQuiz);

        ContentEntryFile updatedFile = new ContentEntryFile();
        updatedFile.setMimeType("application/zip");
        updatedFile.setFileSize(10);
        updatedFile.setLastModified(1540728218);
        updatedFile.setContentEntryFileUid(11);
        contentFileDao.insert(updatedFile);

        ContentEntryContentEntryFileJoin sameFileJoin = new ContentEntryContentEntryFileJoin();
        sameFileJoin.setCecefjContentEntryFileUid(updatedFile.getContentEntryFileUid());
        sameFileJoin.setCecefjContentEntryUid(arabicQuiz.getContentEntryUid());
        sameFileJoin.setCecefjUid(12);
        contentEntryFileJoinDao.insert(sameFileJoin);

        ContentEntryRelatedEntryJoin arabicEnglishJoin = new ContentEntryRelatedEntryJoin();
        arabicEnglishJoin.setCerejContentEntryUid(quiz.getContentEntryUid());
        arabicEnglishJoin.setCerejRelatedEntryUid(arabicQuiz.getContentEntryUid());
        arabicEnglishJoin.setCerejRelLanguageUid(2);
        arabicEnglishJoin.setCerejUid(13);
        arabicEnglishJoin.setRelType(ContentEntryRelatedEntryJoin.REL_TYPE_TRANSLATED_VERSION);
        contentEntryRelatedEntryJoinDao.insert(arabicEnglishJoin);

        ContentEntry spanishQuiz = new ContentEntry();
        spanishQuiz.setContentEntryUid(14);
        spanishQuiz.setTitle("tiempo de prueba");
        spanishQuiz.setThumbnailUrl("https://www.africanstorybook.org/img/asb120.png");
        spanishQuiz.setDescription("todo el contenido");
        spanishQuiz.setPublisher("CK12");
        spanishQuiz.setAuthor("borrachera");
        spanishQuiz.setPrimaryLanguageUid(3);
        spanishQuiz.setLeaf(true);
        contentDao.insert(spanishQuiz);

        ContentEntryFile spanishFile = new ContentEntryFile();
        spanishFile.setMimeType("application/zip");
        spanishFile.setFileSize(10);
        spanishFile.setLastModified(1540728218);
        spanishFile.setContentEntryFileUid(15);
        contentFileDao.insert(spanishFile);

        ContentEntryContentEntryFileJoin spanishFileJoin = new ContentEntryContentEntryFileJoin();
        spanishFileJoin.setCecefjContentEntryFileUid(spanishFile.getContentEntryFileUid());
        spanishFileJoin.setCecefjContentEntryUid(spanishQuiz.getContentEntryUid());
        spanishFileJoin.setCecefjUid(16);
        contentEntryFileJoinDao.insert(spanishFileJoin);

        ContentEntryRelatedEntryJoin spanishEnglishJoin = new ContentEntryRelatedEntryJoin();
        spanishEnglishJoin.setCerejContentEntryUid(quiz.getContentEntryUid());
        spanishEnglishJoin.setCerejRelatedEntryUid(spanishQuiz.getContentEntryUid());
        spanishEnglishJoin.setCerejRelLanguageUid(3);
        spanishEnglishJoin.setCerejUid(17);
        spanishEnglishJoin.setRelType(ContentEntryRelatedEntryJoin.REL_TYPE_TRANSLATED_VERSION);
        contentEntryRelatedEntryJoinDao.insert(spanishEnglishJoin);

        ContentEntryStatus statusStart = new ContentEntryStatus();
        statusStart.setCesLeaf(true);
        statusStart.setDownloadStatus(0);
        statusStart.setTotalSize(1000);
        statusStart.setBytesDownloadSoFar(0);
        statusStart.setCesUid(6);
        statusDao.insert(statusStart);

        ContentEntryStatus statusPending = new ContentEntryStatus();
        statusPending.setCesLeaf(true);
        statusPending.setDownloadStatus(JobStatus.RUNNING);
        statusPending.setTotalSize(1000);
        statusPending.setBytesDownloadSoFar(500);
        statusPending.setCesUid(10);
        statusDao.insert(statusPending);

        ContentEntryStatus statusCompleted = new ContentEntryStatus();
        statusCompleted.setCesLeaf(true);
        statusCompleted.setDownloadStatus(JobStatus.COMPLETE);
        statusCompleted.setTotalSize(1000);
        statusCompleted.setBytesDownloadSoFar(1000);
        statusCompleted.setCesUid(14);
        statusDao.insert(statusCompleted);

        ContentEntryFile file = new ContentEntryFile();
        file.setMimeType("application/webchunk+zip");
        file.setLastModified(System.currentTimeMillis());
        file.setContentEntryFileUid(18);
        fileDao.insert(file);

        ContentEntryContentEntryFileJoin join = new ContentEntryContentEntryFileJoin();
        join.setCecefjContentEntryFileUid(18);
        join.setCecefjContentEntryUid(14);
        join.setCecefjUid(3435);
        contentEntryFileJoinDao.insert(join);

        targetFile = readFromTestResources(
                "/com/ustadmobile/app/android/counting-out-1-20-objects.zip",
                "counting-out-1-20-objects.zip");


        ContentEntryFileStatus fileStatus = new ContentEntryFileStatus();
        fileStatus.setCefsUid(16);
        fileStatus.setCefsContentEntryFileUid(18);
        fileStatus.setFilePath(targetFile.getPath());
        fileStatusDao.insert(fileStatus);


    }

    @Test
    public void givenContentEntryDetailPresent_whenOpened_entryIsDisplayed() throws IOException {
        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(6l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(allOf(withId(R.id.entry_detail_title), withText("Quiz Time")));

        onView(allOf(withId(R.id.entry_detail_description), withText("All content")));

        onView(allOf(withId(R.id.entry_detail_author), withText("Binge")));

    }

    @Test
    public void givenContentEntryDetailPresent_whenTranslatedIsSelected_arabicEntryIsDisplayed() throws IOException {
        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(6l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(Matchers.allOf(isDisplayed(), withId(R.id.entry_detail_flex)))
                .perform(RecyclerViewActions.actionOnItemAtPosition(0, click()));

        intended(AllOf.allOf(
                hasComponent(ContentEntryDetailActivity.class.getCanonicalName()),
                hasExtra(equalTo(ContentEntryListPresenter.ARG_CONTENT_ENTRY_UID),
                        equalTo(String.valueOf(10l))
                )));

        onView(allOf(withId(R.id.entry_detail_title), withText("وقت الاختبار")));

        onView(allOf(withId(R.id.entry_detail_button),
                withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE),
                withText("Download")));


    }

    @Test
    public void givenContentEntryDetailPresent_whenDownloadJobNotStarted_thenShowDownloadButton() throws InterruptedException, IOException {
        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(6l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(withId(R.id.entry_detail_button))
                .check(matches(withText("Download")))
                .check(matches(withEffectiveVisibility(
                        ViewMatchers.Visibility.VISIBLE)));
    }

    @Test
    public void givenContentEntryDetailPresent_whenDownloadJobRunning_thenShowProgressBar() throws IOException {

        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(10l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(withId(R.id.entry_detail_button))
                .check(matches(withEffectiveVisibility(
                        ViewMatchers.Visibility.GONE)));
    }


    @Test
    public void givenContentEntryDetailPresent_whenDownloadJobCompleted_thenCompletedDownloadButton() throws IOException {
        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(14l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(withId(R.id.entry_detail_button))
                .check(matches(withText("Open")))
                .check(matches(withEffectiveVisibility(
                        ViewMatchers.Visibility.VISIBLE)));

    }

    @Test
    public void givenWebChunkContentEntryDetailDownloaded_whenOpenButtonClicked_shouldOpenWebChunkFile() throws IOException, InterruptedException {
        createDummyContent();

        Intent launchActivityIntent = new Intent();
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, String.valueOf(14l));
        mActivityRule.launchActivity(launchActivityIntent);

        onView(allOf(withId(R.id.entry_detail_button), withText("Open")))
              .perform(click());

        Thread.sleep(5000);

        intended(AllOf.allOf(
                hasComponent(WebChunkActivity.class.getCanonicalName()),
                hasExtra(equalTo(WebChunkView.ARG_CHUNK_PATH),
                        equalTo(String.valueOf(targetFile.getPath()))
                )));

    }


}

File: app-android/src/androidTest/java/com/ustadmobile/test/port/android/view/ContentEntryDetailEspressoTest.kt
package com.ustadmobile.test.port.android.view

import android.Manifest
import android.content.Context
import android.content.Intent
import android.view.View
import androidx.recyclerview.widget.RecyclerView
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.contrib.RecyclerViewActions
import androidx.test.espresso.intent.Intents.intended
import androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent
import androidx.test.espresso.intent.matcher.IntentMatchers.hasExtra
import androidx.test.espresso.intent.rule.IntentsTestRule
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.GrantPermissionRule
import androidx.test.runner.AndroidJUnit4
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.ContentEntryDetailPresenter.Companion.ARG_CONTENT_ENTRY_UID
import com.ustadmobile.core.controller.ContentEntryListFragmentPresenter
import com.ustadmobile.core.db.JobStatus
import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.db.dao.ContentEntryStatusDao
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.view.WebChunkView
import com.ustadmobile.lib.db.entities.*
import com.ustadmobile.port.android.generated.MessageIDMap
import com.ustadmobile.port.android.view.ContentEntryDetailActivity
import com.ustadmobile.port.android.view.WebChunkActivity
import org.hamcrest.CoreMatchers.allOf
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.core.AllOf
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException

@RunWith(AndroidJUnit4::class)
class ContentEntryDetailEspressoTest {

    @get:Rule
    var mActivityRule = IntentsTestRule(ContentEntryDetailActivity::class.java, false, false)

    lateinit var context: Context

    @get:Rule
    var permissionRule = GrantPermissionRule.grant(Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.ACCESS_COARSE_LOCATION)


    private var statusDao: ContentEntryStatusDao? = null

    val db: UmAppDatabase
        get() {
            context = InstrumentationRegistry.getInstrumentation().context
            val db = UmAppDatabase.getInstance(context)
            db.clearAllTables()
            return  UmAppDatabase.getInstance(context)// db.getUmRepository("https://localhost", "")
        }

    @Throws(IOException::class)
    fun createDummyContent() {
        val repo = db

        UstadMobileSystemImpl.instance.messageIdMap = MessageIDMap.ID_MAP

        val contentDao = repo.contentEntryDao
        val contentEntryRelatedEntryJoinDao = repo.contentEntryRelatedEntryJoinDao
        val languageDao = repo.languageDao
        val containerDao = repo.containerDao

        val app = UmAppDatabase.getInstance(context)
        statusDao = app.contentEntryStatusDao

        val englishLang = Language()
        englishLang.langUid = 1
        englishLang.name = "English"
        englishLang.iso_639_1_standard = "en"
        englishLang.iso_639_2_standard = "eng"
        englishLang.iso_639_3_standard = "eng"
        languageDao.insert(englishLang)

        val arabicLang = Language()
        arabicLang.langUid = 2
        arabicLang.name = "Arabic"
        arabicLang.iso_639_1_standard = "ar"
        arabicLang.iso_639_2_standard = "ara"
        arabicLang.iso_639_3_standard = "ara"
        languageDao.insert(arabicLang)

        val spanishLang = Language()
        spanishLang.langUid = 3
        spanishLang.name = "Spanish"
        spanishLang.iso_639_1_standard = "es"
        spanishLang.iso_639_2_standard = "esp"
        spanishLang.iso_639_3_standard = "esp"
        languageDao.insert(spanishLang)


        val quiz = ContentEntry()
        quiz.contentEntryUid = 6
        quiz.title = "Quiz Time"
        quiz.thumbnailUrl = "https://www.africanstorybook.org/img/asb120.png"
        quiz.description = "All content"
        quiz.publisher = "CK12"
        quiz.author = "Binge"
        quiz.primaryLanguageUid = 1
        quiz.leaf = true
        contentDao.insert(quiz)

        val contentEntryFile = Container()
        contentEntryFile.mimeType = "application/zip"
        contentEntryFile.fileSize = 10
        contentEntryFile.lastModified = 1540728217
        contentEntryFile.containerContentEntryUid = 8
        containerDao.insert(contentEntryFile)

        // arabic
        val arabicQuiz = ContentEntry()
        arabicQuiz.contentEntryUid = 10
        arabicQuiz.title = "وقت الاختبار"
        arabicQuiz.thumbnailUrl = "https://www.africanstorybook.org/img/asb120.png"
        arabicQuiz.description = "كل المحتوى"
        arabicQuiz.publisher = "CK12"
        arabicQuiz.author = "حفلة"
        arabicQuiz.primaryLanguageUid = 2
        arabicQuiz.leaf = true
        contentDao.insert(arabicQuiz)

        val updatedFile = Container()
        updatedFile.mimeType = "application/zip"
        updatedFile.fileSize = 10
        updatedFile.lastModified = 1540728218
        updatedFile.containerContentEntryUid = 11
        containerDao.insert(updatedFile)

        val arabicEnglishJoin = ContentEntryRelatedEntryJoin()
        arabicEnglishJoin.cerejContentEntryUid = quiz.contentEntryUid
        arabicEnglishJoin.cerejRelatedEntryUid = arabicQuiz.contentEntryUid
        arabicEnglishJoin.cerejRelLanguageUid = 2
        arabicEnglishJoin.cerejUid = 13
        arabicEnglishJoin.relType = ContentEntryRelatedEntryJoin.REL_TYPE_TRANSLATED_VERSION
        contentEntryRelatedEntryJoinDao.insert(arabicEnglishJoin)

        val spanishQuiz = ContentEntry()
        spanishQuiz.contentEntryUid = 14
        spanishQuiz.title = "tiempo de prueba"
        spanishQuiz.thumbnailUrl = "https://www.africanstorybook.org/img/asb120.png"
        spanishQuiz.description = "todo el contenido"
        spanishQuiz.publisher = "CK12"
        spanishQuiz.author = "borrachera"
        spanishQuiz.primaryLanguageUid = 3
        spanishQuiz.leaf = true
        contentDao.insert(spanishQuiz)

        val spanishFile = Container()
        spanishFile.mimeType = "application/zip"
        spanishFile.fileSize = 10
        spanishFile.lastModified = 1540728218
        spanishFile.containerContentEntryUid = 15
        containerDao.insert(spanishFile)

        val spanishEnglishJoin = ContentEntryRelatedEntryJoin()
        spanishEnglishJoin.cerejContentEntryUid = quiz.contentEntryUid
        spanishEnglishJoin.cerejRelatedEntryUid = spanishQuiz.contentEntryUid
        spanishEnglishJoin.cerejRelLanguageUid = 3
        spanishEnglishJoin.cerejUid = 17
        spanishEnglishJoin.relType = ContentEntryRelatedEntryJoin.REL_TYPE_TRANSLATED_VERSION
        contentEntryRelatedEntryJoinDao.insert(spanishEnglishJoin)

        val statusStart = ContentEntryStatus()
        statusStart.cesLeaf = true
        statusStart.downloadStatus = 0
        statusStart.totalSize = 1000
        statusStart.bytesDownloadSoFar = 0
        statusStart.cesUid = 6
        statusDao!!.insert(statusStart)

        val statusPending = ContentEntryStatus()
        statusPending.cesLeaf = true
        statusPending.downloadStatus = JobStatus.RUNNING
        statusPending.totalSize = 1000
        statusPending.bytesDownloadSoFar = 500
        statusPending.cesUid = 10
        statusDao!!.insert(statusPending)

        val statusCompleted = ContentEntryStatus()
        statusCompleted.cesLeaf = true
        statusCompleted.downloadStatus = JobStatus.COMPLETE
        statusCompleted.totalSize = 1000
        statusCompleted.bytesDownloadSoFar = 1000
        statusCompleted.cesUid = 14
        statusDao!!.insert(statusCompleted)

        val file = Container()
        file.mimeType = "application/webchunk+zip"
        file.lastModified = System.currentTimeMillis()
        file.containerContentEntryUid = 18L
        containerDao.insert(file)

    }

    @Test
    @Throws(IOException::class)
    fun givenContentEntryDetailPresent_whenOpened_entryIsDisplayed() {
        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 6L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(allOf<View>(withId(R.id.entry_detail_title), withText("Quiz Time")))

        onView(allOf<View>(withId(R.id.entry_detail_description), withText("All content")))

        onView(allOf<View>(withId(R.id.entry_detail_author), withText("Binge")))

    }

    @Test
    @Throws(IOException::class)
    fun givenContentEntryDetailPresent_whenTranslatedIsSelected_arabicEntryIsDisplayed() {
        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 6L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(allOf<View>(isDisplayed(), withId(R.id.entry_detail_flex)))
                .perform(RecyclerViewActions.actionOnItemAtPosition<RecyclerView.ViewHolder>(0, click()))

        intended(AllOf.allOf(
                hasComponent(ContentEntryDetailActivity::class.java.canonicalName),
                hasExtra(equalTo(ContentEntryListFragmentPresenter.ARG_CONTENT_ENTRY_UID),
                        equalTo(10L.toString())
                )))

        onView(allOf<View>(withId(R.id.entry_detail_title), withText("وقت الاختبار")))

        onView(allOf<View>(withId(R.id.entry_download_open_button),
                withEffectiveVisibility(Visibility.VISIBLE),
                withText("Download")))


    }

    @Test
    @Throws(InterruptedException::class, IOException::class)
    fun givenContentEntryDetailPresent_whenDownloadJobNotStarted_thenShowDownloadButton() {
        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 6L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(withId(R.id.entry_download_open_button))
                .check(matches(withText("Download")))
                .check(matches(withEffectiveVisibility(
                        Visibility.VISIBLE)))
    }

    @Test
    @Throws(IOException::class)
    fun givenContentEntryDetailPresent_whenDownloadJobRunning_thenShowProgressBar() {

        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 10L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(withId(R.id.entry_download_open_button))
                .check(matches(withEffectiveVisibility(
                        Visibility.GONE)))
    }


    @Test
    @Throws(IOException::class)
    fun givenContentEntryDetailPresent_whenDownloadJobCompleted_thenCompletedDownloadButton() {
        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 14L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(withId(R.id.entry_download_open_button))
                .check(matches(withText("Open")))
                .check(matches(withEffectiveVisibility(
                        Visibility.VISIBLE)))

    }

    @Test
    @Throws(IOException::class, InterruptedException::class)
    fun givenWebChunkContentEntryDetailDownloaded_whenOpenButtonClicked_shouldOpenWebChunkFile() {
        createDummyContent()

        val launchActivityIntent = Intent()
        launchActivityIntent.putExtra(ARG_CONTENT_ENTRY_UID, 14L.toString())
        mActivityRule.launchActivity(launchActivityIntent)

        onView(withId(R.id.entry_download_open_button)).perform(click())

        Thread.sleep(5000)

        intended(AllOf.allOf(
                hasComponent(WebChunkActivity::class.java.canonicalName),
                hasExtra(equalTo(WebChunkView.ARG_CONTAINER_UID),
                        equalTo(18L.toString())
                )))

    }


}
